File: backend/src/main/java/com/suplr/backend/service/PdfService.java
package com.suplr.backend.service;

import com.itextpdf.io.font.constants.StandardFonts;
import com.itextpdf.kernel.colors.ColorConstants;
import com.itextpdf.kernel.colors.DeviceRgb;
import com.itextpdf.kernel.font.PdfFont;
import com.itextpdf.kernel.font.PdfFontFactory;
import com.itextpdf.kernel.pdf.PdfDocument;
import com.itextpdf.kernel.pdf.PdfWriter;
import com.itextpdf.layout.Document;
import com.itextpdf.layout.borders.SolidBorder;
import com.itextpdf.layout.element.Cell;
import com.itextpdf.layout.element.Paragraph;
import com.itextpdf.layout.element.Table;
import com.itextpdf.layout.properties.TextAlignment;
import com.itextpdf.layout.properties.UnitValue;
import com.suplr.backend.config.Constants;
import com.suplr.backend.entity.Invoice;
import com.suplr.backend.entity.Order;
import com.suplr.backend.entity.OrderItem;
import com.suplr.backend.entity.Supplier;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;

@Slf4j
@Service
public class PdfService {


    @Value("${app.lbp-rate:90000}")
    private int lbpRate;

    public byte[] renderInvoicePdf(Invoice invoice, Order order, Supplier supplier) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();

        try (PdfWriter writer = new PdfWriter(baos);
             PdfDocument pdf = new PdfDocument(writer);
             Document doc = new Document(pdf)) {

            PdfFont bold = PdfFontFactory.createFont(StandardFonts.HELVETICA_BOLD);
            PdfFont normal = PdfFontFactory.createFont(StandardFonts.HELVETICA);

            Table header = new Table(UnitValue.createPercentArray(new float[]{60, 40}))
                    .useAllAvailableWidth()
                    .setBackgroundColor(Constants.C_DARK)
                    .setPadding(20);

            header.addCell(cell(supplier.getName(), bold, 17, ColorConstants.WHITE)
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            header.addCell(cell("INVOICE", bold, 22, ColorConstants.WHITE)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));

            String contact = supplier.getEmail() != null ? supplier.getEmail() : "";
            if (supplier.getPhone() != null) contact += "  ·  " + supplier.getPhone();
            if (!contact.isBlank()) {
                header.addCell(cell(contact, normal, 8, Constants.C_MUTED)
                        .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            }

            header.addCell(cell(invoice.getNumber(), bold, 11, Constants.C_ACCENT)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            doc.add(header);

            doc.add(new Table(1).useAllAvailableWidth()
                    .setBackgroundColor(Constants.C_ACCENT)
                    .setHeight(4)
                    .addCell(new Cell().setBorder(com.itextpdf.layout.borders.Border.NO_BORDER)));

            String clientName = order.getClient() != null ? order.getClient().getName() : "";
            String clientPhone = order.getClient() != null
                    ? order.getClient().getWhatsappNumber()
                    .replaceAll("@(s\\.whatsapp\\.net|lid)$", "")
                    : "";
            String issued = invoice.getIssuedAt().format(Constants.DATE_FMT);

            Table meta = new Table(UnitValue.createPercentArray(new float[]{48, 4, 48}))
                    .useAllAvailableWidth()
                    .setMarginTop(16);

            Cell billTo = new Cell()
                    .setBackgroundColor(Constants.C_LIGHT)
                    .setPadding(12)
                    .setBorderLeft(new SolidBorder(Constants.C_ACCENT, 3))
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER)
                    .add(new Paragraph("BILL TO").setFont(bold).setFontSize(7).setFontColor(Constants.C_MUTED))
                    .add(new Paragraph(clientName).setFont(bold).setFontSize(12).setFontColor(Constants.C_NAVY))
                    .add(new Paragraph(clientPhone).setFont(normal).setFontSize(9).setFontColor(Constants.C_MUTED));
            meta.addCell(billTo);
            meta.addCell(new Cell().setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));

            Cell metaCard = new Cell()
                    .setBackgroundColor(Constants.C_LIGHT)
                    .setPadding(12)
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .add(new Paragraph("INVOICE DATE").setFont(bold).setFontSize(7).setFontColor(Constants.C_MUTED))
                    .add(new Paragraph(issued).setFont(normal).setFontSize(9).setFontColor(Constants.C_NAVY))
                    .add(new Paragraph("ORDER REFERENCE").setFont(bold).setFontSize(7).setFontColor(Constants.C_MUTED).setMarginTop(8))
                    .add(new Paragraph("#" + order.getId()).setFont(normal).setFontSize(9).setFontColor(Constants.C_NAVY));
            meta.addCell(metaCard);
            doc.add(meta);

            Table items = new Table(UnitValue.createPercentArray(new float[]{46, 16, 20, 18}))
                    .useAllAvailableWidth()
                    .setMarginTop(20);

            for (String h : new String[]{"Description", "Qty", "Unit Price", "Amount"}) {
                items.addHeaderCell(new Cell()
                        .setBackgroundColor(Constants.C_DARK)
                        .add(new Paragraph(h).setFont(bold).setFontSize(8).setFontColor(ColorConstants.WHITE))
                        .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            }

            boolean alt = false;
            for (OrderItem item : order.getItems()) {
                BigDecimal lineTotal = item.getPrice().multiply(item.getQuantity())
                        .setScale(2, RoundingMode.HALF_UP);
                DeviceRgb rowBg = alt ? Constants.C_LIGHT : null;
                alt = !alt;

                addItemCell(items, item.getProductName(), normal, Constants.C_NAVY, rowBg, TextAlignment.LEFT);
                addItemCell(items,
                        item.getQuantity().stripTrailingZeros().toPlainString() + " " + item.getUnit(),
                        normal, Constants.C_MUTED, rowBg, TextAlignment.CENTER);
                addItemCell(items,
                        String.format("%.2f %s", item.getPrice(), invoice.getCurrency()),
                        normal, Constants.C_NAVY, rowBg, TextAlignment.RIGHT);
                addItemCell(items,
                        String.format("%.2f", lineTotal),
                        bold, Constants.C_NAVY, rowBg, TextAlignment.RIGHT);
            }

            items.addCell(new Cell(1, 3)
                    .setBackgroundColor(Constants.C_DARK)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .setPadding(10)
                    .add(new Paragraph("Total").setFont(bold).setFontSize(11).setFontColor(ColorConstants.WHITE))
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            items.addCell(new Cell()
                    .setBackgroundColor(Constants.C_DARK)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .setPadding(10)
                    .add(new Paragraph(String.format("%.2f %s", invoice.getTotal(), invoice.getCurrency()))
                            .setFont(bold).setFontSize(11).setFontColor(ColorConstants.WHITE))
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));

            doc.add(items);

            if ("USD".equals(invoice.getCurrency())) {
                long lbpAmt = invoice.getTotal().longValue() * lbpRate;
                doc.add(new Paragraph(String.format("≈ %,d LBP", lbpAmt))
                        .setFont(normal).setFontSize(8).setFontColor(Constants.C_MUTED)
                        .setTextAlignment(TextAlignment.RIGHT));
            }

            if (order.getNotes() != null && !order.getNotes().isBlank()) {
                Table notesTable = new Table(1).useAllAvailableWidth()
                        .setBackgroundColor(Constants.C_LIGHT)
                        .setMarginTop(16);
                notesTable.addCell(new Cell()
                        .setBorder(new SolidBorder(Constants.C_BORDER, 0.5f))
                        .setPadding(12)
                        .add(new Paragraph("SPECIAL INSTRUCTIONS").setFont(bold).setFontSize(7).setFontColor(Constants.C_MUTED))
                        .add(new Paragraph(order.getNotes().strip()).setFont(normal).setFontSize(9).setFontColor(Constants.C_NAVY)));
                doc.add(notesTable);
            }

            boolean isPaid = invoice.getPaidAt() != null;
            DeviceRgb badgeBg = isPaid ? Constants.C_PAID_BG : Constants.C_DUE_BG;
            DeviceRgb badgeFg = isPaid ? Constants.C_PAID_FG : Constants.C_DUE_FG;
            String badgeTxt = isPaid ? "PAID IN FULL" : "PAYMENT OUTSTANDING";

            Table footer = new Table(UnitValue.createPercentArray(new float[]{30, 70}))
                    .useAllAvailableWidth()
                    .setMarginTop(24)
                    .setBorder(new SolidBorder(Constants.C_BORDER, 0.5f));
            footer.addCell(new Cell()
                    .setBackgroundColor(badgeBg)
                    .setPadding(12)
                    .setTextAlignment(TextAlignment.CENTER)
                    .add(new Paragraph(badgeTxt).setFont(bold).setFontSize(9).setFontColor(badgeFg))
                    .setBorderRight(new SolidBorder(Constants.C_BORDER, 0.5f))
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            footer.addCell(new Cell()
                    .setPadding(12)
                    .setTextAlignment(TextAlignment.RIGHT)
                    .add(new Paragraph("Thank you for your business, " + clientName + "!")
                            .setFont(normal).setFontSize(9).setFontColor(Constants.C_MUTED))
                    .add(new Paragraph("Generated with Suplr")
                            .setFont(normal).setFontSize(7).setFontColor(Constants.C_MUTED))
                    .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER));
            doc.add(footer);

        } catch (IOException e) {
            log.error("PDF generation failed for invoice {}: {}", invoice.getNumber(), e.getMessage());
            throw new RuntimeException("PDF generation failed", e);
        }

        return baos.toByteArray();
    }

    private static DeviceRgb rgb(int r, int g, int b) {
        return new DeviceRgb(r, g, b);
    }

    private static Cell cell(String text, PdfFont font, float size, com.itextpdf.kernel.colors.Color color) {
        return new Cell().add(new Paragraph(text).setFont(font).setFontSize(size).setFontColor(color))
                .setBorder(com.itextpdf.layout.borders.Border.NO_BORDER);
    }

    private static void addItemCell(Table table, String text, PdfFont font,
                                    DeviceRgb color, DeviceRgb bg, TextAlignment align) {
        Cell c = new Cell()
                .add(new Paragraph(text).setFont(font).setFontSize((float) 9).setFontColor(color))
                .setTextAlignment(align)
                .setPadding(8)
                .setBorderBottom(new SolidBorder(Constants.C_BORDER, 0.4f))
                .setBorderLeft(com.itextpdf.layout.borders.Border.NO_BORDER)
                .setBorderRight(com.itextpdf.layout.borders.Border.NO_BORDER)
                .setBorderTop(com.itextpdf.layout.borders.Border.NO_BORDER);
        if (bg != null) c.setBackgroundColor(bg);
        table.addCell(c);
    }
}

File: backend/src/pdf/pdf.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import PDFDocument from 'pdfkit';
import type { Invoice } from '../entities/invoice.entity.js';
import type { Order } from '../entities/order.entity.js';
import type { Supplier } from '../entities/supplier.entity.js';

const C_DARK = '#0F172A';
const C_ACCENT = '#10B981';
const C_MUTED = '#64748B';
const C_LIGHT = '#F8FAFC';
const C_NAVY = '#1E293B';
const C_BORDER = '#E2E8F0';
const C_WHITE = '#FFFFFF';
const C_PAID_BG = '#DCFCE7';
const C_PAID_FG = '#166534';
const C_DUE_BG = '#FEF3C7';
const C_DUE_FG = '#92400E';

@Injectable()
export class PdfService {
  private readonly logger = new Logger(PdfService.name);
  private readonly lbpRate: number;

  constructor(config: ConfigService) {
    this.lbpRate = Number(config.get('LBP_RATE', '90000'));
  }

  renderInvoicePdf(invoice: Invoice, order: Order, supplier: Supplier): Buffer {
    const chunks: Buffer[] = [];
    const doc = new PDFDocument({ margin: 40, size: 'A4' });
    doc.on('data', (c: Buffer) => chunks.push(c));

    const W = 515;
    const left = 40;

    doc.rect(left - 40, 0, 595, 80).fill(C_DARK);
    doc
      .fillColor(C_WHITE)
      .font('Helvetica-Bold')
      .fontSize(17)
      .text(supplier.name, left, 20, { width: W * 0.6 });
    doc
      .font('Helvetica-Bold')
      .fontSize(22)
      .fillColor(C_WHITE)
      .text('INVOICE', left, 20, { width: W, align: 'right' });

    const contact = [supplier.email, supplier.phone]
      .filter(Boolean)
      .join('  ·  ');
    if (contact) {
      doc
        .font('Helvetica')
        .fontSize(8)
        .fillColor(C_MUTED)
        .text(contact, left, 44, { width: W * 0.6 });
    }
    doc
      .font('Helvetica-Bold')
      .fontSize(11)
      .fillColor(C_ACCENT)
      .text(invoice.number, left, 44, { width: W, align: 'right' });

    doc.rect(left - 40, 80, 595, 4).fill(C_ACCENT);

    const metaY = 100;
    const halfW = (W - 10) / 2;
    const clientName = order.client?.name ?? '';
    const clientPhone = (order.client?.whatsappNumber ?? '').replace(
      /@(s\.whatsapp\.net|lid)$/,
      '',
    );
    const issued = invoice.issuedAt.toLocaleDateString('en-US', {
      month: 'long',
      day: '2-digit',
      year: 'numeric',
    });

    doc.rect(left, metaY, halfW, 70).fill(C_LIGHT);
    doc.rect(left, metaY, 3, 70).fill(C_ACCENT);
    doc
      .font('Helvetica-Bold')
      .fontSize(7)
      .fillColor(C_MUTED)
      .text('BILL TO', left + 12, metaY + 10);
    doc
      .font('Helvetica-Bold')
      .fontSize(12)
      .fillColor(C_NAVY)
      .text(clientName, left + 12, metaY + 22, { width: halfW - 15 });
    doc
      .font('Helvetica')
      .fontSize(9)
      .fillColor(C_MUTED)
      .text(clientPhone, left + 12, metaY + 40, { width: halfW - 15 });

    const metaX = left + halfW + 10;
    doc.rect(metaX, metaY, halfW, 70).fill(C_LIGHT);
    doc
      .font('Helvetica-Bold')
      .fontSize(7)
      .fillColor(C_MUTED)
      .text('INVOICE DATE', metaX, metaY + 10, {
        width: halfW,
        align: 'right',
      });
    doc
      .font('Helvetica')
      .fontSize(9)
      .fillColor(C_NAVY)
      .text(issued, metaX, metaY + 20, { width: halfW, align: 'right' });
    doc
      .font('Helvetica-Bold')
      .fontSize(7)
      .fillColor(C_MUTED)
      .text('ORDER REFERENCE', metaX, metaY + 36, {
        width: halfW,
        align: 'right',
      });
    doc
      .font('Helvetica')
      .fontSize(9)
      .fillColor(C_NAVY)
      .text(`#${order.id}`, metaX, metaY + 46, {
        width: halfW,
        align: 'right',
      });

    let y = metaY + 82;
    const cols = [
      left,
      left + W * 0.46,
      left + W * 0.62,
      left + W * 0.82,
      left + W,
    ];
    const colWidths = [W * 0.46, W * 0.16, W * 0.2, W * 0.18];
    const headers = ['Description', 'Qty', 'Unit Price', 'Amount'];
    const hAligns: ('left' | 'center' | 'right')[] = [
      'left',
      'center',
      'right',
      'right',
    ];

    doc.rect(left, y, W, 22).fill(C_DARK);
    headers.forEach((h, i) => {
      doc
        .font('Helvetica-Bold')
        .fontSize(8)
        .fillColor(C_WHITE)
        .text(h, cols[i] + 4, y + 7, {
          width: colWidths[i] - 8,
          align: hAligns[i],
        });
    });
    y += 22;

    let alt = false;
    for (const item of order.items) {
      const lineTotal = (
        parseFloat(item.price) * parseFloat(item.quantity)
      ).toFixed(2);
      if (alt) doc.rect(left, y, W, 24).fill(C_LIGHT);
      doc
        .moveTo(left, y + 24)
        .lineTo(left + W, y + 24)
        .strokeColor(C_BORDER)
        .lineWidth(0.4)
        .stroke();
      doc
        .font('Helvetica')
        .fontSize(9)
        .fillColor(C_NAVY)
        .text(item.productName, cols[0] + 4, y + 8, {
          width: colWidths[0] - 8,
          align: 'left',
        });
      doc
        .font('Helvetica')
        .fontSize(9)
        .fillColor(C_MUTED)
        .text(`${parseFloat(item.quantity)} ${item.unit}`, cols[1] + 4, y + 8, {
          width: colWidths[1] - 8,
          align: 'center',
        });
      doc
        .font('Helvetica')
        .fontSize(9)
        .fillColor(C_NAVY)
        .text(
          `${parseFloat(item.price).toFixed(2)} ${invoice.currency}`,
          cols[2] + 4,
          y + 8,
          { width: colWidths[2] - 8, align: 'right' },
        );
      doc
        .font('Helvetica-Bold')
        .fontSize(9)
        .fillColor(C_NAVY)
        .text(lineTotal, cols[3] + 4, y + 8, {
          width: colWidths[3] - 8,
          align: 'right',
        });
      y += 24;
      alt = !alt;
    }

    doc.rect(left, y, W, 32).fill(C_DARK);
    doc
      .font('Helvetica-Bold')
      .fontSize(11)
      .fillColor(C_WHITE)
      .text('Total', left + 4, y + 10, { width: W * 0.82 - 8, align: 'right' });
    doc
      .font('Helvetica-Bold')
      .fontSize(11)
      .fillColor(C_WHITE)
      .text(
        `${parseFloat(invoice.total).toFixed(2)} ${invoice.currency}`,
        cols[3] + 4,
        y + 10,
        { width: colWidths[3] - 8, align: 'right' },
      );
    y += 32;

    if (invoice.currency === 'USD') {
      const lbpAmt = Math.floor(parseFloat(invoice.total) * this.lbpRate);
      doc
        .font('Helvetica')
        .fontSize(8)
        .fillColor(C_MUTED)
        .text(`≈ ${lbpAmt.toLocaleString()} LBP`, left, y + 4, {
          width: W,
          align: 'right',
        });
      y += 18;
    }

    if (order.notes?.trim()) {
      y += 12;
      doc.rect(left, y, W, 50).fill(C_LIGHT);
      doc.rect(left, y, W, 50).strokeColor(C_BORDER).lineWidth(0.5).stroke();
      doc
        .font('Helvetica-Bold')
        .fontSize(7)
        .fillColor(C_MUTED)
        .text('SPECIAL INSTRUCTIONS', left + 12, y + 10);
      doc
        .font('Helvetica')
        .fontSize(9)
        .fillColor(C_NAVY)
        .text(order.notes.trim(), left + 12, y + 22, { width: W - 24 });
      y += 62;
    }

    y += 16;
    const isPaid = !!invoice.paidAt;
    const badgeBg = isPaid ? C_PAID_BG : C_DUE_BG;
    const badgeFg = isPaid ? C_PAID_FG : C_DUE_FG;
    const badgeTxt = isPaid ? 'PAID IN FULL' : 'PAYMENT OUTSTANDING';

    doc.rect(left, y, W, 44).strokeColor(C_BORDER).lineWidth(0.5).stroke();
    doc.rect(left, y, W * 0.3, 44).fill(badgeBg);
    doc
      .font('Helvetica-Bold')
      .fontSize(9)
      .fillColor(badgeFg)
      .text(badgeTxt, left + 4, y + 18, {
        width: W * 0.3 - 8,
        align: 'center',
      });
    doc
      .font('Helvetica')
      .fontSize(9)
      .fillColor(C_MUTED)
      .text(
        `Thank you for your business, ${clientName}!`,
        left + W * 0.3 + 8,
        y + 10,
        { width: W * 0.7 - 12, align: 'right' },
      );
    doc
      .font('Helvetica')
      .fontSize(7)
      .fillColor(C_MUTED)
      .text('Generated with Suplr', left + W * 0.3 + 8, y + 26, {
        width: W * 0.7 - 12,
        align: 'right',
      });

    doc.end();
    return Buffer.concat(chunks);
  }
}
